### DISCUSSION TITLE
How to Allow Users to Download Multiple Variants in Saleor?

### DISCUSSION THREAD

**Discussion by daltonsena** [saleor/saleor#13822] [General] [answered]
Hello everyone,

First, I've extensively searched through open and closed issues, consulted the documentation and discussions, and used demo.saleor.io. Since I couldn't find what I was looking for in these channels, I am opening this issue to find a solution.

I am trying to create a workflow where a buyer can access multiple variants of the same product (I am not sure if "variants" is the most appropriate term here; please enlighten me). In other words, the user can download the versions they prefer after making a purchase.

Let's consider a hypothetical example. Suppose that:

- Product Type (Audiobook):
    - Is this product shippable? (No)
    - Regular product type (Yes)
    - Product type uses variant attributes (Yes)
        - Variants
            - WAV
            - MP3

When my users make a purchase, I would like them to have the option to download any variant they want.

I don't know the correct approach, or even if this is entirely against Saleor's rules or impossible. Should I give up?

I am open to some customization; would developing a plugin be helpful?

Some points to consider:

- Would it be wiser to use Product Attributes instead of Variants?
- How can I use the Digital Content structure in this context?
- I am relatively familiar with the metadata API, but I'm not keen on storing links via metadata because these links need to be pre-signed.
- In my scenario, the dashboard is not very relevant. I plan to create, manipulate, and upload these products at a low level (API, scripting, etc....) if necessary.

**Thank you for your time and consideration!**

----

> System information
> Saleor version: latest
> Saleor storefront: latest
> Saleor dashboard: latest
> 
> Operating system:
> MacOS
>

**Comment by daltonsena** [comment]
Has anyone seen this? any ideas here?

**Comment by timuric** [comment, marked as answer]
You would have to create custom logic in your storefront that would allow users to download files based on desired qualifications. Probably it is not necessary to extend saleor at all, it can be adjacent functionality of your storefront

**Reply by bandieramonte** [reply]
Thank you @timuric 
I would like to know whether the Saleor API already has the capability to handle multiple product downloads per variant chose of the same product, as exposed in the 1st post here. 
And then, when the user chooses a certain variant, does the API offer the ability to the user to log in into its account, access the downloadable content, and download the product the user just paid for? 
Then also can the API handle the possibility for a manager in the dashboard to upload digital content per variant, which the user later downloads?
If this all is possible with the API, then it's just a matter of adjusting the front end to make this happen.
Kindly let me know.
Thank you!

**Reply by timuric** [reply]
Our API allows all of the scenarios you described. What you are describing are front end features that Saleor has no opinion on. Admin wise you can upload files per variant.

**Reply by bandieramonte** [reply]
@timuric regarding uploading files per variant admin wise, we are already discussing this [here](https://github.com/saleor/saleor/discussions/9482#discussioncomment-8708949)

Regarding the API being able to accomplish all the scenarios I described, I am not able to find enough documentation in Saleor's website about this. For instance, I search "digital content" in your docs, and come across this [link](https://docs.saleor.io/search?q=digital%20content). There I am not able to find how the API allows for all the scenarios I described, for example, to allow the user to download a certain content after having paid for it. Can you kindly point me to the API documentation describing these scenarios? Thank you

**Reply by timuric** [reply]
@bandieramonte, keep in mind that the product catalog and attributes are public queries by default. Are you exposing Saleor API to the world? If so then you would need to move files to authorized storage outside of saleor

**Reply by bandieramonte** [reply]
@timuric yes, thank you for letting me know of this.
btw, I found [this link](https://saleor-fork.readthedocs.io/en/latest/architecture/products.html#digital-products) with a bit more info about handling digital content with Saleor's API. Is this still valid? Anything else you might have about this topic would be helpful for me to expand on this functionality. Thank you.
